agent/credentialssource/credentials_source.go

```diff
@@ -0,0 +1,196 @@
+	"github.com/percona/pmm/api/inventorypb"
+)
+
+// Parser is a struct which is responsible for parsing credentialsJSON source/defaults file.
+type Parser struct{}
+
+// New creates new Parser.
+func New() *Parser {
+	return &Parser{}
+}
+
+type credentialsSource struct {
+	username      string
+	password      string
+	host          string
+	agentPassword string
+	port          uint32
+	socket        string
+}
+
+// credentialsJSON provides access to an external provider so that
+// the username, password, or agent password can be managed
+// externally, e.g. HashiCorp Vault, Ansible Vault, etc.
+type credentialsJSON struct {
+	AgentPassword string `json:"agentpassword"`
+	Password      string `json:"password"`
+	Username      string `json:"username"`
+}
+
+// ParseCredentialsSource parses given file in request. It returns the database specs.
+func (d *Parser) ParseCredentialsSource(req *agentpb.ParseCredentialsSourceRequest) *agentpb.ParseCredentialsSourceResponse {
+	var res agentpb.ParseCredentialsSourceResponse
+	parsedData, err := parseCredentialsSourceFile(req.FilePath, req.ServiceType)
+	if err != nil {
+		res.Error = err.Error()
+		return &res
+	}
+
+	res.Username = parsedData.username
+	res.Password = parsedData.password
+	res.AgentPassword = parsedData.agentPassword
+	res.Host = parsedData.host
+	res.Port = parsedData.port
+	res.Socket = parsedData.socket
+
+	return &res
+}
+
+func parseCredentialsSourceFile(filePath string, serviceType inventorypb.ServiceType) (*credentialsSource, error) {
+	if len(filePath) == 0 {
+		return nil, errors.New("configPath for parseCredentialsSourceFile is empty")
+	}
+
+	filePath, err := expandPath(filePath)
+	if err != nil {
+		return nil, fmt.Errorf("fail to normalize path: %w", err)
+	}
+
+	// check if file exist
+	if _, err := os.Stat(filePath); errors.Is(err, os.ErrNotExist) {
+		return nil, errors.Errorf("file doesn't exist: %s", filePath)
+	}
+
+	credentialsJSONFile, err := parseJsonFile(filePath)
+	if err == nil {
+		return credentialsJSONFile, nil
+	}
+
+	if serviceType == inventorypb.ServiceType_MYSQL_SERVICE {
+		return parseMySQLDefaultsFile(filePath)
+	}
+
+	return nil, errors.Errorf("unrecognized file type %s", filePath)
+}
+
+func parseJsonFile(filePath string) (*credentialsSource, error) {
+	creds := credentialsJSON{"", "", ""}
+
+	f, err := os.Lstat(filePath)
+	if err != nil {
+		return nil, fmt.Errorf("%w", err)
+	}
+
+	if f.Mode()&0o111 != 0 {
+		return nil, fmt.Errorf("%w: %s", errors.New("file execution is not supported"), filePath)
+	}
+
+	// Read the file
+	content, err := readFile(filePath)
+	if err != nil {
+		return nil, fmt.Errorf("%w", err)
+	}
+
+	if err := json.Unmarshal([]byte(content), &creds); err != nil {
+		return nil, fmt.Errorf("%w", err)
+	}
+
+	parsedData := &credentialsSource{
+		username:      creds.Username,
+		password:      creds.Password,
+		agentPassword: creds.AgentPassword,
+	}
+
+	err = validateResults(parsedData)
+	if err != nil {
+		return nil, err
+	}
```

Comment:
if we validate data in both functions can we do it in https://github.com/percona/pmm/pull/1110/files#diff-3f486b304421a573b79690cddeb7fe149e9d4159999e8ad5bad717bbd4d85bc9R60 once we get the result?